lib/tests/test_ssh_signing.rs

```diff
@@ -0,0 +1,164 @@
+use std::io::Write;
+use std::process::Stdio;
+
+use jj_lib::signing::{SigStatus, SigningBackend};
+use jj_lib::ssh_signing::SshBackend;
+
+static PRIVATE_KEY: &str = r#"-----BEGIN OPENSSH PRIVATE KEY-----
+b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
+QyNTUxOQAAACBo/iejekjvuD/HTman0daImstssYYR52oB+dmr1KsOYQAAAIiuGFMFrhhT
+BQAAAAtzc2gtZWQyNTUxOQAAACBo/iejekjvuD/HTman0daImstssYYR52oB+dmr1KsOYQ
+AAAECcUtn/J/jk/+D5+/+WbQRNN4eInj5L60pt6FioP0nQfGj+J6N6SO+4P8dOZqfR1oia
+y2yxhhHnagH52avUqw5hAAAAAAECAwQF
+-----END OPENSSH PRIVATE KEY-----
+"#;
+
+static PUBLIC_KEY: &str =
+    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGj+J6N6SO+4P8dOZqfR1oiay2yxhhHnagH52avUqw5h";
+
+struct SshEnvironment {
+    private_key: Option<tempfile::TempPath>,
+    allowed_signers: Option<tempfile::TempPath>,
+}
+
+impl SshEnvironment {
+    fn new() -> Result<Self, std::process::Output> {
+        let mut private_key = tempfile::Builder::new()
+            .prefix("jj-test-pk-")
+            .tempfile()
+            .unwrap();
+
+        private_key.write_all(PRIVATE_KEY.as_bytes()).unwrap();
+        private_key.flush().unwrap();
+
+        let private_key_path = private_key.into_temp_path();
+
+        let ssh_add = std::process::Command::new("ssh-add")
```

Comment:
Does it mean the key is added to my ssh-agent if I run `cargo test` locally? Can we somehow set up isolated environment?

Reply:
I don't know of a way to set up an isolated environment with ssh-agent. I don't think it is possible but I might be wrong.

The ssh key is removed when the test exits and the backing temp file is removed (any keys added on ssh-agent won't work without the backing file on disk).

I don't think it would ever pollute the host in a way that would result in the key being accidentally used. But yea this is not ideal.

Reply:
> I don't know of a way to set up an isolated environment with ssh-agent. I don't think it is possible but I might be wrong.

I think we should just start another `ssh-agent` instance and make sure we add the key to that instance (and connect to that instance for reading the key, of course).

Reply:
Hmm I was under the impression that all agents would share the same data but I was wrong about that. I tried this out and it actually works pretty well. I have updated the test implementation to spawn a dedicated agent for the test.

Reply:
Do we need the agent at all? Can we somehow teach `ssh` to use the test key file by using environment variable or config file?

Reply:
> Do we need the agent at all?

The doc says the `-f` option "may refer to either a private key, or a public key with the private half available via ssh-agent(1)." So `backend.sign(data, Some(PRIVATE_KEY))` removes the need of the ssh agent.

@julienvincent does it make sense to test the `SshBackend` API that way?

BTW, I don't think the user would want to inline SSH public key in `~/.jjconfig.toml`. Maybe better to change `sign()` API to take `key` as a file path, like in Git?
https://docs.gitlab.com/ee/user/project/repository/signed_commits/ssh.html#configure-git-to-sign-commits-with-your-ssh-key

Reply:
I originally wanted to make use of the ssh-agent to test the full pathway as that is most likely how this will be used out in the wild. That being said we probably shouldn't be writing tests for ssh-keygen and can just assume it works as advertised haha.

I updated the tests to work without ssh-agent.

Regarding the inlining of the SSH public key - git does also allow this inlining and is how I prefer having it setup. We should definitely support both options though so I have updated the implementation to work with either a file path or an inlined pub key to align with git.

> BTW, I don't think the user would want to inline SSH public key in `~/.jjconfig.toml`.

While you're right that git allows specifying a file-path here, git also allows inlining the pub key as I have done here. This is actually how I currently have my git signing configured and how I personally prefer to use it.

I imagine we would want to support both options similar to git - I updated the implementation to allow passing either a file-path or an inlined public key.